### PR TITLE
initial floating ip association commit

### LINKED ISSUE #7: Add floating ip association resource
Moved from https://github.com/hetznercloud/terraform-provider-hcloud/issues/33

### PR DESCRIPTION
**Pull request by arminbuerkle**
Offers a solution for creating a floating IP before the server and attaching it once the server is ready.

For reference: https://github.com/hetznercloud/terraform-provider-hcloud/issues/33

Closes #7

### CODE REVIEW

**Comment by arminbuerkle** [comment]
@thetechnick didn't have much time the last days but here are finally the 2 PRs from the previous repo updated.

There still a bug in here which i couldn't quite figure out: 

When creating a floating_ip_association it shows the server and floating ip values as 0: 
```hcl
hcloud_floating_ip_association.test[0]
      id:             <computed>
      floating_ip_id: "0"
      server_id:      "0"
```

The result on remote and in the terraform.tfstate is correct though.

Similar when manually unassigning a server from a floating ip. A following `terraform plan` will show: 
```hcl
      server_id:      "123456" => "0"
``` 
A `terraform apply` will do the correct thing however.

**Comment by volym3ad** [comment]
@thetechnick can you please review it?
This functionality is kinda necessary for me right now.

**Comment by LKDevelopment** [comment]
@arminbuerkle would it be possible that you gave me an access to your repo?

**Comment by arminbuerkle** [comment]
@LKDevelopment somehow your comment about access was lost. I just added you as a collaborator to the project.

File: hcloud/resource_hcloud_floatingip_association.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+			},
+		},
+	}
+}
+
+func resourceFloatingIPAssociationCreate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*hcloud.Client)
+	ctx := context.Background()
+
+	floatingIPID, ok := d.GetOk("floating_ip_id")
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: `floating_ip_id` is already required, no need to check it's set.

File: hcloud/resource_hcloud_floatingip_association.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	if !ok {
+		return fmt.Errorf("could not find floating ip id")
+	}
+	floatingIP := &hcloud.FloatingIP{ID: floatingIPID.(int)}
+
+	serverID, ok := d.GetOk("server_id")
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: `server_id` is already required, no need to check it's set.

File: hcloud/resource_hcloud_floatingip_association.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	return resourceFloatingIPAssociationRead(d, m)
+}
+
+func resourceFloatingIPAssociationRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*hcloud.Client)
+	ctx := context.Background()
+
+	_, err := strconv.Atoi(d.Id())
+	if err != nil {
+		log.Printf("[WARN] Floating IP Association ID (%s) not found, removing from state: %v", d.Id(), err)
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: Please change to `Floating IP ID (%s)`

File: hcloud/resource_hcloud_floatingip_association.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	// check if correct server is associated
+	if floatingIP.Server != nil {
+		server.ID = floatingIP.Server.ID
+	} else {
+		log.Printf("[WARN] Floating IP (%v) is not associated to a server, removing Floating IP Association from state", d.Get("floating_ip_id"))
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("server_id", server.ID)
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: use `floatingIP.Server.ID` here and negate the `if` on line 95

File: hcloud/resource_hcloud_floatingip_association.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"floating_ip_id": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"server_id": {
+				Type:     schema.TypeInt,
+				Required: true,
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: We should mark this as `ForceNew` and remove the Update function

File: hcloud/resource_hcloud_floatingip_association.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	return resourceFloatingIPAssociationRead(d, m)
+}
+
+func resourceFloatingIPAssociationDelete(d *schema.ResourceData, m interface{}) error {
+	client := m.(*hcloud.Client)
+	ctx := context.Background()
+
+	_, err := strconv.Atoi(d.Id())
+	if err != nil {
+		log.Printf("[WARN] Floating IP Association ID (%s) not found, removing from state: %v", d.Id(), err)
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: This is not a `not found` error, but something along `invalid id` (similar to other resources)

File: hcloud/resource_hcloud_floatingip_association.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+		return nil
+	}
+
+	floatingIPID, ok := d.GetOk("floating_ip_id")
+	if !ok {
+		log.Printf("[WARN] Floating IP ID (%v) not found, removing Floating IP Association from state", d.Get("floating_ip_id"))
+		d.SetId("")
+		return nil
+	}
+	floatingIP := &hcloud.FloatingIP{ID: floatingIPID.(int)}
```
Comment (thetechnick) [inline hcloud/resource_hcloud_floatingip_association.go]: Please change this to use the record id and not `floating_ip_id`.
So we do no longer have to validate `floating_ip_id`

**Comment by namxam** [comment]
As there seems to be progress on this… any news on when to expect it to be available? This is quite critical for our setup and it would be nice if it makes sense to wait or if we have to find a "hack" to get it working.

**Comment by LKaemmerling** [comment]
@namxam we plan to provide a new version within this or the next week.

File: CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,7 @@
 FEATURES:
 
 * **New Resource**: `hcloud_rnds` Manage reverse DNS entries for servers and Floating IPs.
+* **New Resource**: `hcloud_floating_ip_association` Manage the association between floating ips and servers.
```
Comment (thcyron) [inline CHANGELOG.md]: floating ips → Floating IPs

File: website/docs/r/floating_ip_association.html.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+resource "hcloud_floating_ip" "master" {
+  type = "ipv4"
+  home_location = "nbg1"
+}
+```
+
+## Argument Reference
+
+- `floating_ip_id` - (Required) Id of the Floating IP.
```
Comment (thcyron) [inline website/docs/r/floating_ip_association.html.md]: Id → ID

File: website/docs/r/floating_ip_association.html.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+- `server_id` - (Required) Server to assign the Floating IP to.
+
+## Attributes Reference
+
+- `id` - Unique ID of the Floating IP Association.
+- `floating_ip_id` - Id of the Floating IP.
```
Comment (thcyron) [inline website/docs/r/floating_ip_association.html.md]: Id → ID

File: hcloud/resource_hcloud_floatingip_association.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+	floatingIPID, err := strconv.Atoi(d.Id())
+	if err != nil {
+		log.Printf("[WARN] Invalid id (%s), removing from state: %v", d.Id(), err)
+		d.SetId("")
+		return nil
+	}
+
+	floatingIP := &hcloud.FloatingIP{ID: floatingIPID}
+
+	_, _, err = client.FloatingIP.Unassign(ctx, floatingIP)
```
Comment (thcyron) [inline hcloud/resource_hcloud_floatingip_association.go]: The API returns an error should the Floating IP already be unassigned. I think we should not treat that as an error.

Unfortunately, the API doesn’t return a dedicated error code in that case. I can take care of that to make the API return an error code à la `floating_ip_already_unassigned`.

File: CHANGELOG.md (hunk 8 above)
Comment (thcyron) [inline CHANGELOG.md]: In the API we talk about assigning and unassigning Floating IPs. What about naming the resource `hcloud_floating_ip_assignment` then to be consistent?

**Review by thcyron — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 8 above)
Comment (LKaemmerling) [reply CHANGELOG.md]: This sounds good. I will update everything.

**Pull request merged**